File: persistence/src/main/java/com/comssa/persistence/question/major/admin/dto/RequestMakeMultipleChoiceQuestionDto.java
package com.comssa.persistence.question.major.admin.dto;

import com.comssa.persistence.question.common.domain.Question;
import com.comssa.persistence.question.common.domain.QuestionChoice;
import com.comssa.persistence.question.common.dto.request.RequestMakeQuestionDto;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

import java.util.List;
import java.util.stream.Collectors;

@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
@Getter
public class RequestMakeMultipleChoiceQuestionDto extends RequestMakeQuestionDto {
	private List<RequestMakeQuestionChoiceDto> questionChoices;

	public static <T extends QuestionChoice> RequestMakeMultipleChoiceQuestionDto from(
		Question question,
		List<T> questionChoices) {
		return RequestMakeMultipleChoiceQuestionDto.builder()
			.content(question.getContent())
			.description(question.getDescription())
			.questionCategory(question.getQuestionCategory())
			.questionLevel(question.getQuestionLevel())
			.questionChoices(questionChoices
				.stream()
				.map(RequestMakeQuestionChoiceDto::from)
				.collect(Collectors.toList()))
			.build();
	}
}

File: persistence/src/main/java/com/comssa/persistence/question/dto/common/request/RequestMakeMultipleChoiceQuestionDto.kt
package com.comssa.persistence.question.dto.common.request

import com.comssa.persistence.question.domain.common.Question
import com.comssa.persistence.question.domain.common.QuestionCategory
import com.comssa.persistence.question.domain.common.QuestionChoice
import com.comssa.persistence.question.domain.common.QuestionLevel

class RequestMakeMultipleChoiceQuestionDto(
	val questionChoices: List<RequestMakeQuestionChoiceDto>,
	content: String,
	questionCategory: QuestionCategory,
	questionLevel: QuestionLevel,
	description: String,
) : RequestMakeQuestionDto(
		content = content,
		questionCategory = questionCategory,
		questionLevel = questionLevel,
		description = description,
	) {
	companion object {
		@JvmStatic
		fun <T : QuestionChoice> from(
			question: Question,
			questionChoices: List<T>,
		): RequestMakeMultipleChoiceQuestionDto {
			val mappedChoices =
				questionChoices.map { questionChoice ->
					RequestMakeQuestionChoiceDto.from(questionChoice)
				}
			return RequestMakeMultipleChoiceQuestionDto(
				questionChoices = mappedChoices,
				content = question.content,
				description = question.description,
				questionCategory = question.questionCategory,
				questionLevel = question.questionLevel,
			)
		}
	}
}
